File: MCP服务器/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/go/pkg/crypto/crypto.go
// Package crypto provides ECDSA key management and signature operations for SchemaPin.
package crypto

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"crypto/x509"
	"encoding/asn1"
	"encoding/base64"
	"encoding/pem"
	"fmt"
	"math/big"
)

// KeyManager handles ECDSA key operations
type KeyManager struct{}

// NewKeyManager creates a new KeyManager instance
func NewKeyManager() *KeyManager {
	return &KeyManager{}
}

// GenerateKeypair generates a new ECDSA key pair using P-256 curve
func (k *KeyManager) GenerateKeypair() (*ecdsa.PrivateKey, error) {
	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("failed to generate ECDSA key pair: %w", err)
	}
	return privateKey, nil
}

// ExportPrivateKeyPEM exports private key to PEM format using PKCS#8
func (k *KeyManager) ExportPrivateKeyPEM(key *ecdsa.PrivateKey) (string, error) {
	keyBytes, err := x509.MarshalPKCS8PrivateKey(key)
	if err != nil {
		return "", fmt.Errorf("failed to marshal private key: %w", err)
	}

	block := &pem.Block{
		Type:  "PRIVATE KEY",
		Bytes: keyBytes,
	}

	return string(pem.EncodeToMemory(block)), nil
}

// ExportPublicKeyPEM exports public key to PEM format
func (k *KeyManager) ExportPublicKeyPEM(key *ecdsa.PublicKey) (string, error) {
	keyBytes, err := x509.MarshalPKIXPublicKey(key)
	if err != nil {
		return "", fmt.Errorf("failed to marshal public key: %w", err)
	}

	block := &pem.Block{
		Type:  "PUBLIC KEY",
		Bytes: keyBytes,
	}

	return string(pem.EncodeToMemory(block)), nil
}

// LoadPrivateKeyPEM loads private key from PEM format
func (k *KeyManager) LoadPrivateKeyPEM(pemData string) (*ecdsa.PrivateKey, error) {
	block, _ := pem.Decode([]byte(pemData))
	if block == nil {
		return nil, fmt.Errorf("failed to decode PEM block")
	}

	// Try PKCS#8 format first (preferred)
	if key, err := x509.ParsePKCS8PrivateKey(block.Bytes); err == nil {
		if ecdsaKey, ok := key.(*ecdsa.PrivateKey); ok {
			return ecdsaKey, nil
		}
		return nil, fmt.Errorf("not an ECDSA private key")
	}

	// Fall back to EC private key format
	key, err := x509.ParseECPrivateKey(block.Bytes)
	if err != nil {
		return nil, fmt.Errorf("failed to parse private key: %w", err)
	}

	return key, nil
}

// LoadPublicKeyPEM loads public key from PEM format
func (k *KeyManager) LoadPublicKeyPEM(pemData string) (*ecdsa.PublicKey, error) {
	block, _ := pem.Decode([]byte(pemData))
	if block == nil {
		return nil, fmt.Errorf("failed to decode PEM block")
	}

	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, fmt.Errorf("failed to parse public key: %w", err)
	}

	ecdsaKey, ok := pub.(*ecdsa.PublicKey)
	if !ok {
		return nil, fmt.Errorf("not an ECDSA public key")
	}

	return ecdsaKey, nil
}

// CalculateKeyFingerprint computes SHA-256 fingerprint of public key
func (k *KeyManager) CalculateKeyFingerprint(key *ecdsa.PublicKey) (string, error) {
	keyBytes, err := x509.MarshalPKIXPublicKey(key)
	if err != nil {
		return "", fmt.Errorf("failed to marshal public key for fingerprint: %w", err)
	}

	hash := sha256.Sum256(keyBytes)
	return fmt.Sprintf("sha256:%x", hash), nil
}

// CalculateKeyFingerprintFromPEM computes SHA-256 fingerprint from PEM-encoded public key
func (k *KeyManager) CalculateKeyFingerprintFromPEM(publicKeyPEM string) (string, error) {
	publicKey, err := k.LoadPublicKeyPEM(publicKeyPEM)
	if err != nil {
		return "", err
	}
	return k.CalculateKeyFingerprint(publicKey)
}

// SignatureManager handles signature operations
type SignatureManager struct{}

// NewSignatureManager creates a new SignatureManager instance
func NewSignatureManager() *SignatureManager {
	return &SignatureManager{}
}

// ecdsaSignature represents the ASN.1 structure for ECDSA signatures
type ecdsaSignature struct {
	R, S *big.Int
}

// SignHash signs a hash with the private key and returns base64-encoded signature
func (s *SignatureManager) SignHash(hashBytes []byte, privateKey *ecdsa.PrivateKey) (string, error) {
	r, sig, err := ecdsa.Sign(rand.Reader, privateKey, hashBytes)
	if err != nil {
		return "", fmt.Errorf("failed to sign hash: %w", err)
	}

	// Encode signature in ASN.1 DER format for cross-language compatibility
	signature := ecdsaSignature{R: r, S: sig}
	derBytes, err := asn1.Marshal(signature)
	if err != nil {
		return "", fmt.Errorf("failed to marshal signature: %w", err)
	}

	return base64.StdEncoding.EncodeToString(derBytes), nil
}

// VerifySignature verifies a base64-encoded signature against a hash
func (s *SignatureManager) VerifySignature(hashBytes []byte, signatureB64 string, publicKey *ecdsa.PublicKey) bool {
	signature, err := base64.StdEncoding.DecodeString(signatureB64)
	if err != nil {
		return false
	}

	// Decode ASN.1 DER signature
	var sig ecdsaSignature
	if _, err := asn1.Unmarshal(signature, &sig); err != nil {
		return false
	}

	return ecdsa.Verify(publicKey, hashBytes, sig.R, sig.S)
}

// SignSchemaHash signs a schema hash (convenience method)
func (s *SignatureManager) SignSchemaHash(schemaHash []byte, privateKey *ecdsa.PrivateKey) (string, error) {
	return s.SignHash(schemaHash, privateKey)
}

// VerifySchemaSignature verifies a schema signature (convenience method)
func (s *SignatureManager) VerifySchemaSignature(schemaHash []byte, signatureB64 string, publicKey *ecdsa.PublicKey) bool {
	return s.VerifySignature(schemaHash, signatureB64, publicKey)
}

File: reference-mcp-servers/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/python/schemapin/crypto.py
"""Cryptographic operations for SchemaPin using ECDSA P-256."""

import base64
import hashlib
from typing import Tuple

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.asymmetric.ec import (
    EllipticCurvePrivateKey,
    EllipticCurvePublicKey,
)


class KeyManager:
    """Manages ECDSA P-256 key generation and serialization."""

    @staticmethod
    def generate_keypair() -> Tuple[EllipticCurvePrivateKey, EllipticCurvePublicKey]:
        """
        Generate new ECDSA P-256 key pair.

        Returns:
            Tuple of (private_key, public_key)
        """
        private_key = ec.generate_private_key(ec.SECP256R1())
        public_key = private_key.public_key()
        return private_key, public_key

    @staticmethod
    def export_private_key_pem(private_key: EllipticCurvePrivateKey) -> str:
        """
        Export private key to PEM format.

        Args:
            private_key: ECDSA private key

        Returns:
            PEM-encoded private key string
        """
        pem_bytes = private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.PKCS8,
            encryption_algorithm=serialization.NoEncryption()
        )
        return pem_bytes.decode('utf-8')

    @staticmethod
    def export_public_key_pem(public_key: EllipticCurvePublicKey) -> str:
        """
        Export public key to PEM format.

        Args:
            public_key: ECDSA public key

        Returns:
            PEM-encoded public key string
        """
        pem_bytes = public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        )
        return pem_bytes.decode('utf-8')

    @staticmethod
    def load_private_key_pem(pem_data: str) -> EllipticCurvePrivateKey:
        """
        Load private key from PEM format.

        Args:
            pem_data: PEM-encoded private key string

        Returns:
            ECDSA private key
        """
        return serialization.load_pem_private_key(
            pem_data.encode('utf-8'),
            password=None
        )

    @staticmethod
    def load_public_key_pem(pem_data: str) -> EllipticCurvePublicKey:
        """
        Load public key from PEM format.

        Args:
            pem_data: PEM-encoded public key string

        Returns:
            ECDSA public key
        """
        return serialization.load_pem_public_key(pem_data.encode('utf-8'))

    @staticmethod
    def calculate_key_fingerprint(public_key: EllipticCurvePublicKey) -> str:
        """
        Calculate SHA-256 fingerprint of public key.

        Args:
            public_key: ECDSA public key

        Returns:
            SHA-256 fingerprint in format 'sha256:hexstring'
        """
        der_bytes = public_key.public_bytes(
            encoding=serialization.Encoding.DER,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        )
        fingerprint = hashlib.sha256(der_bytes).hexdigest()
        return f"sha256:{fingerprint}"

    @staticmethod
    def calculate_key_fingerprint_from_pem(public_key_pem: str) -> str:
        """
        Calculate SHA-256 fingerprint from PEM-encoded public key.

        Args:
            public_key_pem: PEM-encoded public key string

        Returns:
            SHA-256 fingerprint in format 'sha256:hexstring'
        """
        public_key = KeyManager.load_public_key_pem(public_key_pem)
        return KeyManager.calculate_key_fingerprint(public_key)


class SignatureManager:
    """Manages ECDSA signature creation and verification."""

    @staticmethod
    def sign_hash(hash_bytes: bytes, private_key: EllipticCurvePrivateKey) -> str:
        """
        Sign hash using ECDSA P-256 and return Base64-encoded signature.

        Args:
            hash_bytes: SHA-256 hash to sign
            private_key: ECDSA private key

        Returns:
            Base64-encoded signature
        """
        signature = private_key.sign(hash_bytes, ec.ECDSA(hashes.SHA256()))
        return base64.b64encode(signature).decode('ascii')

    @staticmethod
    def verify_signature(hash_bytes: bytes, signature_b64: str, public_key: EllipticCurvePublicKey) -> bool:
        """
        Verify ECDSA signature against hash.

        Args:
            hash_bytes: Original SHA-256 hash
            signature_b64: Base64-encoded signature
            public_key: ECDSA public key

        Returns:
            True if signature is valid, False otherwise
        """
        try:
            signature = base64.b64decode(signature_b64)
            public_key.verify(signature, hash_bytes, ec.ECDSA(hashes.SHA256()))
            return True
        except Exception:
            return False

    @classmethod
    def sign_schema_hash(cls, schema_hash: bytes, private_key: EllipticCurvePrivateKey) -> str:
        """
        Sign schema hash and return Base64 signature.

        Args:
            schema_hash: SHA-256 hash of canonical schema
            private_key: ECDSA private key

        Returns:
            Base64-encoded signature
        """
        return cls.sign_hash(schema_hash, private_key)

    @classmethod
    def verify_schema_signature(
        cls,
        schema_hash: bytes,
        signature_b64: str,
        public_key: EllipticCurvePublicKey
    ) -> bool:
        """
        Verify schema signature against hash.

        Args:
            schema_hash: SHA-256 hash of canonical schema
            signature_b64: Base64-encoded signature
            public_key: ECDSA public key

        Returns:
            True if signature is valid, False otherwise
        """
        return cls.verify_signature(schema_hash, signature_b64, public_key)
